designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+in contrast to 'major changes', which are by their nature semantic, and require
+human-in-the-loop reasoning to accommodate. Major changes, such the interface
+changes to the differential equation solvers, may be announced via deprecation
+warnings in minor versions, but will not be implemented/removed prior to a
+version 3.0.0.
+
+A minor feature must be deprecated for a minimum of 3 minor versions
+(approximately 1 year on the current release schedule) before removal. The exact
+timeline for a feature's removal will be decided during the development which
+would mark it as deprecated.
+
+A minor feature is deprecated and a candidate for future removal when:
+1. stanc emits a compile-time warning with the following 3 parts:
+    1. A description of the deprecated feature, such as "Use of <- for assignment
+       is deprecated".
+    2. An _action_ the user should take, such as "Use = instead". If stanc3 can
+       automate this for them, we should call that out, leading to a full
+       message like "Use = instead or run stanc3 --print-canonical to update
+       print an updated version of your code."
+    3. A timeline which includes a date or version after which this feature is
+       slated for removal, for example "This deprecation will expire in Stan
+       version 2.31.0 (expected Sept. 2022)".
+2. It has been added to the [documentation on
+   deprecation](https://mc-stan.org/docs/reference-manual/deprecated-features-appendix.html)
+   with the same 3 pieces of information, and additional information as needed
+   (such as any of: reasoning for removal, advantages of new behavior, ongoing
+   related changes, etc).
+3. Examples in the documentation have been updated to remove any examples of the
+   deprecated syntax, except for those referring to its deprecation. This change
+   may lag behind bullets 1 and 2, but must occur before the removal.
+
+Our goal is to only deprecate a feature when doing so provides a tangible
+benefit to users, such as improving error messaging or allowing new features to
+be added to the language while keeping a low maintenance burden. Changes should
+never take users (even those with many or large models) long to adjust to, and
+as few such changes should be made in each version as is reasonable.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Developers
+From a developer standpoint, this policy is rather straightforward. When a
+change is made, it should identify any deprecation candidates: new keywords that
+must be implemented (thus 'deprecating' the use of that word as a identifier),
+new syntax which supersedes prior syntax, (such as `=` or `array[]`), or a
+similar **syntax-level** change. If supporting both the new and old way of doing
+things requires a 'hack' or difficult to maintain code, this is a prime
+candidate for this deprecation-removal process. If a change meaningfully changes
+the semantics of the language, it is not.
```

Comment:
I think we can have arbitrarily compatibility-breaking deprecations.   So maybe rewrite this doc to explicitly introduce major deprecations (anything that's not minor).  We don't want to say it's impossible, just that it's not going to auto-deprecate in three releases.  We'll have to wait for major changes.  

Do we want to wait at least a year after deprecation before changing major version in the same way?

Reply:
I think it's all about expectations. My instinct is major versions would not require the same timeline, but that is partially because I expect a new major version would be announced pretty far in advance anyway.